File: Projekt Fehlerfindung Programmcode/EvaluationFramework/adts/stack.py
from adts import ADT, ADTMethod, ADTSignature
from adts.domains import CharDomain, IntDomain, BoolDomain


# Stack ADT implementation with a limited static capacity
class Stack(ADT):
    # ADT consists of a constructor and methods for pushing, popping and checking whether the stack is full or empty
    @staticmethod
    def get_signature() -> ADTSignature:
        # Sample chars from the alphabet and the integer 3
        char_dom = CharDomain()
        int_dom = IntDomain(3, 3)
        bool_dom = BoolDomain()

        method_list = [
            ADTMethod("create", [int_dom], None, False),
            ADTMethod("push", [char_dom], None, False),
            ADTMethod("pop", None, char_dom, False),
            ADTMethod("is_full", None, bool_dom, True),
            ADTMethod("is_empty", None, bool_dom, True)
        ]

        return {
            method.name: method for method in method_list
        }

    # Remember the capacity limit for later and create a list for element storage
    def __init__(self, capacity: int) -> None:
        self.capacity: int = capacity
        self.content: list[str] = []

    # Add the value on top of the stack if the capacity has not been reached yet, otherwise raise an error
    def push(self, value: str) -> None:
        if len(self.content) >= self.capacity:
            raise RuntimeError("Cannot push to a full stack")

        self.content.append(value)

    # Return the value from the top of the stack, remove it and raise an error if the stack is empty
    def pop(self) -> str:
        if len(self.content) == 0:
            raise RuntimeError("Cannot pop from an empty stack")

        return self.content.pop()

    # Check whether the stack capacity has been reached
    def is_full(self) -> bool:
        return len(self.content) >= self.capacity

    # Check whether there are currently no elements on the stack
    def is_empty(self) -> bool:
        return len(self.content) == 0

File: Projekt Fehlerfindung Programmcode/EvaluationFramework/assets/stack.pl
:- initialization(main, main).

anysize(0).
anysize(1).
anysize(2).
anysize(3).
anysize(4).

anystack(stack_(push_(S, _), N)) :- anysize(N), anystack(stack_(S, N)).
anystack(stack_(create_, N)) :- anysize(N).

size(create_, 0).
size(push_(S, _), N) :- size(S, K), N is K + 1, !.

create(C, stack_(create_, C)) :- anysize(C).
create(C, stack_(_, C)) :- ab(create), anysize(C).

push(stack_(_, N1), _, error) :- ab(push), anysize(N1).
push(stack_(_, N1), _, stack_(_, N2)) :- ab(push), anysize(N1), anysize(N2).
push(stack_(S, C), E, stack_(push_(S, E), C)) :- anysize(N), anysize(C), size(S, N), N < C.
push(stack_(S, C), _, error) :- anysize(N), anysize(C), size(S, N), N >= C.

pop(stack_(_, N1), _, error) :- ab(pop), anysize(N1).
pop(stack_(_, N1), _, stack_(_, N2)) :- ab(pop), anysize(N1), anysize(N2).
pop(stack_(push_(S, E), C), E, stack_(S, C)) :- anysize(C).
pop(stack_(create_, C), error, error) :- anysize(C).

isempty(stack_(_, N), _) :- ab(isempty), anysize(N).
isempty(stack_(create_, C), true) :- anysize(C).
isempty(stack_(push_(_, _), C), false) :- anysize(C).

isfull(stack_(_, N), _) :- ab(isfull), anysize(N).
isfull(stack_(S, C), true) :- size(S, N), anysize(N), anysize(C), N >= C.
isfull(stack_(S, C), false) :- size(S, N), anysize(N), anysize(C), N < C.
